lib/utils/note-utils.js

```diff
@@ -10,7 +10,7 @@ import { isFunction } from 'lodash';
  *
  * @type {RegExp} matches a title and excerpt in note content
  */
-const noteTitleAndPreviewRegExp = /\s*(\S[^\n]*)\s*(\S[^\n]*)?/;
+const noteTitleAndPreviewRegExp = /\s*(\S[^\n]*)\s*([\s\S]*)?/;
```

Comment:
This is a curious change here. What are you intending to communicate with the `([\s\S]*)` group? To me that reads the same as `(.*)`

Reply:
`.` does not include line terminators such as `/n`.  I thought the exact same thing!

Reply:
https://regex101.com/r/Gtac3V/1

Reply:
As an historical note the single first line I think was chosen for pragmatic reasons when considering performance. We don't want a ten-thousand line note to slow down the note list.

If we want to chop these at a specific limit we can build that into the RegExp as well or feed a chopped version of the input string into the RegExp.

This may be a good time to improve the overall quality of this algorithm if indeed we're going to touch it. In reviewing what we want we could consider something like this…

```js
const noteTitleAndPreviewRegExp = /\s*([^\n]{1,32})[^\n]*\n+[ \t]*([\s\S]{1,300})/;
```

Put another way…

```js
const noteTitleAndPreviewRegExp =new RexExp(
  '\s*' + // ignore leading whitespace
  '([^\n]{1,32})' + // then grab up to so many title characters (on a single line)
  '[^\n]*\n+' + // then skip the remainder of the title characters and any following newlines
  '[ \t]*' + // skip the leading whitespace
  '([\s\S]{1,300})' // and grab the first 300 characters of the note
);

…

const [ , title, rawPreview ] = match;
const preview = rawPreview.split('\n').slice(0,4).join('\n'); // the RegExp allows for 300 one-character lines
```

I'm trying to do this all in one RegExp: allow up to four lines maxing out at a total number of characters. I'll update if I figure it out

Reply:
Here are some thoughts on behaviors…
 - we want a max length of the title
 - we want a max length of the preview characters
 - we also want a max number of lines in the preview characters
 - we want to skip over blank lines when making the preview

```js
const getTitleAndPreview = content => {
  const titlePattern = /\s*([^\n]{1,10})/g; // grab the first N=10 characters of the first line
  const titleMatch = titlePattern.exec( content );
  if ( ! titleMatch ) {
    return [ 'Untitled', '' ]; // do something appropriate here
  }
  const [ , title ] = titleMatch;

  const previewPattern = /[^\n]*\n+[ \t]*([^\n]{0,10})/g; // skip leading whitespace, then grab first N=10 characters of body
  previewPattern.lastIndex = titlePattern.lastIndex; // start where the title ended

  const go = ( pattern, source, preview = '', lines = 0 ) => {
    const match = pattern.exec( source );
    // must match, must have no more than four lines, must not be longer than N=10 characters
    if ( ! match || lines > 3 || preview.length > 10 ) {
      return preview;
    }

    const [ , chunk ] = match;
    return go( pattern, source, preview + chunk, lines + 1 );
  }

  return [ title, go( previewPattern, content ) ];
}
```

there's cleanup work here but it addresses one thing I don't like about the existing solution or my previous comment: the logic is self-contained and doesn't rely on tricks

this looks complicated and can probably be simplified quite a bit but I think it produces our exactly-desired behaviors and does so without running into performance issues for massive notes. the `N=10` thing is controllable by us

Reply:
```js
const maxTitleChars = 64;
const maxPreviewChars = 200;
const maxPreviewLines = 4; // probably need to adjust if we're in the comfy view

const matchUntilLimit = ( pattern, source, preview = '', lines = 0 ) => {
    const match = pattern.exec( source );
    // must match, must have no more than four lines, must not be longer than N=10 characters
    if ( ! match || lines > maxPreviewLines || preview.length > maxPreviewChars ) {
      return preview;
    }

    const [ , chunk ] = match;
    return matchUntilLimit( pattern, source, preview + chunk, lines + 1 );
  }

const getTitleAndPreview = content => {
  const titlePattern = new RegExp( `\s*([^\n]{1,${ maxTitleChars }})`, 'g' ); // grab the first N=10 characters of the first line
  const titleMatch = titlePattern.exec( content );
  if ( ! titleMatch ) {
    return [ 'Untitled', '' ]; // do something appropriate here
  }
  const [ , title ] = titleMatch;

  const previewPattern = new RegExp( `[^\n]*\n+[ \t]*([^\n]{0,${ maxPreviewChars }})`, 'g' ); // skip leading whitespace, then grab first N=10 characters of body
  previewPattern.lastIndex = titlePattern.lastIndex; // start where the title ended

  return [ title, matchUntilLimit( previewPattern, content ) ];
}
```

Reply:
Your version:
```
The/long Note!.
Call to getTitleAndPreview took 0.2849999873433262 milliseconds.
```
My version:
```
The/long Note!.
Call to getTitleAndPreview took 0.9950000094249845 milliseconds.
```

Reply:
That being said, if we identified a performance issue then I think a big one is under our noses: I think we calculate this on ever render of the sidebar, when indeed we only need run it on our initial index and then whenever note content changes.

We could store this in a separate state value and eliminate any performance issues altogether. Work for a future iteration.